### ISSUE TITLE
Application maximum poll interval (300000ms) exceeded by 88msApplication maximum poll interval (300000ms) exceeded by 88ms

### ISSUE THREAD

**Issue by ybbiubiubiu** [confluentinc/confluent-kafka-python#552]
Description
===========

When I use subprocess.Popen in a flask project to open a script (the script instantiates the consumer object) to pull the message (using api consume and poll), when the consumer pulls a part of the data, it hangs. I will wait until 60000ms to report this error.

Application maximum poll interval (300000ms) exceeded by 88ms(adjust max.poll.interval.ms for long-running message processing): leaving group

Please help me to see how to solve it.

How to reproduce
================
Here are some of the code blocks in my script.

-----------------------------
 ```
consumer = Consumer({'bootstrap.servers':"xxx:xxx.xxx.xxx:9092",
                         "group.id":"xxx",
                         # 'enable.auto.commit':False,
                         "fetch.wait.max.ms":3000,
                         # 'fetch.min.bytes':1024 * 512,
                         'session.timeout.ms':6000,
                         "on_commit":_on_send_response,
                         "default.topic.config":{"auto.offset.reset": "earliest"}
                        }
                         )

    pt = TopicPartition("data_checksum2", 0)  
    consumer.assign([pt])
   try:
        while True:
            filter_dict2 = filter_dict
            # time.sleep()
            msg = consumer.poll(0.2)
            if msg:
                table = json.loads(msg.value().decode("utf-8")).get('table')
                if table in filter_dict2:
                    # gtid = json.loads((msg.value().decode("utf-8"))).get('gtid')
                    # offset = msg.offset()
                    print(json.loads((msg.value().decode("utf-8"))).get('es'), )
                    try:
                        producer.produce(str(filter_dict2.get(table)), msg.value())
                    except Exception:
                        producer.poll(0)
                        producer.flush(1)
                elif msg.error():
                    if msg.error().code() == KafkaError._PARTITION_EOF:
                        continue
                    else:
                        aelog.error(msg.error())
            else:
                continue
    except Exception as e:
        print(e)
    finally:
        try:
            aelog.info("同步提交补救成功")
            print(consumer.committed([pt]))
        except KafkaError as e:
            consumer.close()
```

Checklist
=========
Please provide the following information:

 - [0.11.6 ] confluent-kafka-python and librdkafka version (`confluent_kafka.version()` and `confluent_kafka.libversion()`):
 - [2.1.0 ] Apache Kafka broker version:
 - [ ] Client configuration: `{...}`
 - [ ] Operating system:
 - [ ] Provide client logs (with `'debug': '..'` as necessary)
 - [ ] Provide broker log excerpts
 - [ ] Critical issue

MAXPOLL|rdkafka#consumer-1| [thrd:main]: Application maximum poll interval (300000ms) exceeded by 88ms (adjust max.poll.interval.ms for long-running message processing): leaving group

**Comment by mhowlett** [comment]
the reference to `max.poll.interval.ms` implies you're using librdkafka version 1.0 (or a custom compiled version from master after 0.11.6), not 0.11.6. is that correct? 

`max.poll.interval.ms` is an important parameter for applications where processing of messages can potentially take a long time (introduced in 1.0). previously, this was effectively set to the session.timeout.ms (which you ideally want to have set relatively low).

the error message you're seeing means you waited longer than `max.poll.interval.ms` between calls to consumer.poll. it's not immediately clear to me from your code / explanation how that is happening. does `aelog.error(msg.error())` block?

**Comment by edenhill** [comment]
Do make sure that you are creating the client instances (producer, consumer) in the process you aim to use them, a client instance WILL NOT be usable in a forked child-process due to the background threads not surviving the fork barrier.

**Comment by ybbiubiubiu** [comment]
> the reference to `max.poll.interval.ms` implies you're using librdkafka version 1.0 (or a custom compiled version from master after 0.11.6), not 0.11.6. is that correct?
> 
> `max.poll.interval.ms` is an important parameter for applications where processing of messages can potentially take a long time (introduced in 1.0). previously, this was effectively set to the session.timeout.ms (which you ideally want to have set relatively low).
> 
> the error message you're seeing means you waited longer than `max.poll.interval.ms` between calls to consumer.poll. it's not immediately clear to me from your code / explanation how that is happening. does `aelog.error(msg.error())` block?

thanks ，i get it。

**Comment by june6868** [comment]
so,how to fix it?

**Comment by deepaksood619** [comment]
I am not able to catch this exception... How to catch this exception?

```
while True:
    msg = c.poll(1.0)

    # initial error handling
    if msg is None:
        continue

    if msg.error():
        if msg.error().code() == KafkaError._PARTITION_EOF:
            continue
        else:
            logging.error(f'consumer error: {msg.error()}')
            break
```

Error is not caught in `logging.error`, the consumer leaves the group and never recovers and nor exits.

**Comment by edenhill** [comment]
Remove the `break` from the error case, the client will automatically recover and rejoin the group as soon as you call poll() again.

**Comment by deepaksood619** [comment]
@edenhill Thanks for a quick response.

The problem is that i don't think the control of execution is coming to `break`, since if `break` is called the program will exit and kubernetes will restart the container.

Also the error log i am getting is 
`%4|1562783637.645|MAXPOLL|rdkafka#consumer-1| [thrd:main]: Application maximum poll interval (300000ms) exceeded by 398ms (adjust max.poll.interval.ms for long-running message processing): leaving group` 
and the statement to print the error is `logging.error(f'consumer error: {msg.error()}')`, so i don't think the error is printed using the print statement i wrote. I am not able to understand from where this `error` is printed in my code.

Here is my whole code just for reference -
https://gist.github.com/deepaksood619/b41d65baf26601118a6b9294b806e60e

Any help regarding how can i improve this or how can i debug this will be helpful.

Thanks in advance.

**Comment by edenhill** [comment]
The MAXPOLL error will be logged if consumer.poll() is not called at least every max.poll.interval.ms; I'm noticing some backoff-and-retry sleeps in your http code, is it possible that these kicked in for longer than 30s when this happened?

**Comment by deepaksood619** [comment]
Yes, this is what is happening. How can i make my consumer robust, so that if `leaving group` it should exit. I am not able to catch this exception in my code. I want to catch this exception if thread is busy in http call.

On a different note, how should i monitor consumer lag in prometheus/grafana. I am not able to get consumer lag metrics via `prometheus-jmx-exporter` from kafka. I am currently using kafkamanager to see the lag but i want to get that metrics in prometheus.

**Comment by edenhill** [comment]
It is not exception, it is a log message, and it can't and shouldn't be catched.
The max.poll.interval.ms is there for a reason; it let's you specify how long your consumer owns the assigned partitions following a rebalance - doing anything with the data when this period has expired means there might be duplicate processing.
So the solution is to either:
 a. indicate that your application is still alive by calling poll() - if you dont want more messages you will need to pause() your partitions first (but do note that this comes at the cost of purging the pre-fetch queue).
 b. increase max.poll.interval.ms to your maximum http retry time.

**Comment by deepaksood619** [comment]
Thanks a lot.. Now i understand a lot better.

Can a consumer rejoin a consumer group after it has left the group?

Also any tips regarding monitoring consumer lag?

**Comment by edenhill** [comment]
The consumer will rejoin as soon as you call poll() again.

confluent-control-center allows you to monitor consumer lag.

**Comment by deepaksood619** [comment]
@edenhill Thank you so much.

**Comment by geoff-va** [comment]
I'm running into a similar situation where I'm waiting to commit the offsets until after I've done some processing on the pulled data. 

I'm pulling, say, 2M values via a loop of `poll()`, then once I've reached a certain offset for each partition, I `pause` that partition. both my partitions are paused and I head off to process my data and insert it into a db. Once that's successful I commit the offsets.

It takes some time to process the data, so by the time I get around to committing the offsets I see the following:

```raised unexpected: KafkaException('KafkaError{code=ILLEGAL_GENERATION,val=22,str="Commit failed: Broker: Specified group generation id is not valid"}',)```

Based on the above, it sounded like as long as the consumer was paused then this shouldn't be an issue? Is there sometime else I need to do to deal with this? 

Currently Using `confluent-kafka==1.0.1`

**Comment by geoff-va** [comment]
I am also seeing this occur:
```cimpl.KafkaException: KafkaError{code=UNKNOWN_MEMBER_ID,val=25,str="Commit failed: Broker: Unknown member"}```

when calling: `consumer.commit(asynchronous=False)`

**Comment by PrasadPingle** [comment]
Hi @ybbiubiubiu how do resolved this issue? I'm also facing the same issue. Will appreciate any help on this.

**Comment by rogaha** [comment]
same here: https://github.com/confluentinc/confluent-kafka-go/issues/344#issuecomment-719876972

**Comment by edenhill** [comment]
You will need to call poll() at least every max.poll.interval.ms, regardless if you've paused the partitions or not.

**Comment by antoniodimariano** [comment]
> It is not exception, it is a log message, and it can't and shouldn't be catched.
> The max.poll.interval.ms is there for a reason; it let's you specify how long your consumer owns the assigned partitions following a rebalance - doing anything with the data when this period has expired means there might be duplicate processing.
> So the solution is to either:
> a. indicate that your application is still alive by calling poll() - if you dont want more messages you will need to pause() your partitions first (but do note that this comes at the cost of purging the pre-fetch queue).
> b. increase max.poll.interval.ms to your maximum http retry time.

I call poll() in a while True loop but I got the same error, 
```
|MAXPOLL|rdkafka#consumer-3| [thrd:main]: Application maximum poll interval (300000ms) exceeded by 183ms (adjust max.poll.interval.ms for long-running message processing): leaving group

```
soon after the consumer started

The main login of my consumer is 
```
 def run(self):


        try:
            self.consumer.subscribe(self.consumer_topic, on_assign=self.print_assignment)
            while True:
                incoming_message = self.consumer.poll(timeout=1.0)

                if incoming_message is None:
                    continue
                elif incoming_message.error() is None:
                         process_message(incoming_message.value())                        

                elif incoming_message.error():
                    if incoming_message.error().code() == KafkaError._PARTITION_EOF:
                        sys.stderr.write('%% %s [%d] reached end at offset %d\n' %
                                         (incoming_message.topic(), incoming_message.partition(),
                                          incoming_message.offset()))
                    elif incoming_message.error():
                        raise KafkaException(incoming_message.error())

        except Exception as whatever_it_is:
            pass

        finally:
            self.consumer.close()
           
            os._exit(-1)

```
I am using the confluent-kafka 1.5.0

**Comment by edenhill** [comment]
How long does your process_message() call take?

Tip: consumer errors are generally to be treated as informational as the consumer will try to recover automatically, so the re-raising of the consumer error may be removed and replaced with an error log instead.

**Comment by antoniodimariano** [comment]
I usually use `self.executor.submit(process_message, payload=incoming_message.value())` and it takes ~1sec  to complete. It simply add the message' payload to a dict.

**Comment by edenhill** [comment]
Recommend timing the process_message() calls to see if this is what takes time, or something else.

**Comment by antoniodimariano** [comment]
Ok. I wondered why the consumer doesn't rejoin the group? Let's say the process_message() takes time for some internal problem. Why I end up with a not working consumer if the Application maximum poll interval is exceeded?

**Comment by edenhill** [comment]
It will rejoin the group the next time you call poll(), but your loop exits (re-raises) the exception and closes the consumer.

Consumer errors are generally to be considered informational as the consumer will attempt to recover automatically, so I recommend changing your `raise ..` to a log and a continue.

We're planning to extend the error exceptions with a flag saying if it is a permanent or temporary error which will allow your application to take appropriate actions (e.g., exit on permanent, ignore on temporary).

**Comment by antoniodimariano** [comment]
> It will rejoin the group the next time you call poll(), but your loop exits (re-raises) the exception and closes the consumer.
> 
> Consumer errors are generally to be considered informational as the consumer will attempt to recover automatically, so I recommend changing your `raise ..` to a log and a continue.
> 
> We're planning to extend the error exceptions with a flag saying if it is a permanent or temporary error which will allow your application to take appropriate actions (e.g., exit on permanent, ignore on temporary).

@edenhill thanks, this will be a nice one to have. 


I am also trying to recreate it and get the `Application maximum poll interval leaving group` message back.
I am setting the  `session.timeout.ms': 6000 so to make it happens faster. 
This is the consumer 

```

from confluent_kafka import TopicPartition, KafkaException, KafkaError, OFFSET_END
from confluent_kafka.avro import AvroConsumer
import sys
import time

class Consumer:

    def __init__(self, **kwargs):
        consumer_conf = {'session.timeout.ms': 6000,
                'enable.auto.commit': 'true',
                'log.connection.close': 'false',
                'default.topic.config': {'auto.offset.reset': 'latest'},
                'schema.registry.url': 'https://xxxxxxx:8081',
                'bootstrap.servers': 'xxxxxx:9091',
                'group.id': 'my_unique_consumer_id'}
        self.consumer = AvroConsumer(consumer_conf)
        self.consumer_topic = kwargs.get('topic')

    def process_message(self,message):
        print("long process")
        time.sleep(30)
        print("end process")


    def run(self):

       
        try:
            self.consumer.subscribe(self.consumer_topic, on_assign=self.print_assignment)

            while True:

                incoming_message = self.consumer.poll(1)
                if incoming_message is None:
                    continue
                elif incoming_message.error() is None:
                    sys.stderr.write(
                        '[Topic:%s][Message:%s] \n' % (
                            str.upper(incoming_message.topic()), incoming_message.value()))
                    self.process_message(incoming_message.value())

                elif incoming_message.error():
                    if incoming_message.error().code() == KafkaError._PARTITION_EOF:
                        # End of partition event
                        sys.stderr.write('%% %s [%d] reached end at offset %d\n' %
                                         (incoming_message.topic(), incoming_message.partition(),
                                          incoming_message.offset()))
                    elif incoming_message.error():
                        sys.stderr.write(
                            'Consumer Error {incoming_message.error()}\n')
                        raise KafkaException(incoming_message.error())


        except Exception as error:

            sys.stderr.write(
                ' EXCEPTION %s on the consumer for topic(s) %s\n' % (
                    error, self.consumer_topic))


        finally:
            sys.stderr.write(
                'CLOSING consumer %s with topic(s): %s\n' % (
                    self.consumer, self.consumer_topic))
            self.consumer.close()
            return 1

```
I am keeping the `raise KafkaException(incoming_message.error())` in the code and using a blocking function `process_message ` to process the message. So the consumer remained alive even if the `process_message` took 30secs and the 'session.timeout.ms': 6000 and I didn't get any error. 

```
long process
%7|1610626629.638|HEARTBEAT|rdkafka#consumer-1| [thrd:main]: GroupCoordinator/8: Heartbeat for group "my_consumer_unique_id" generation id 9
%7|1610626629.675|OFFSET|rdkafka#consumer-1| [thrd:main]: Topic my_test_topic [0]: stored offset 25, committed offset 24: setting stored offset 25 for commit
%7|1610626630.024|COMMIT|rdkafka#consumer-1| [thrd:main]: GroupCoordinator/8: OffsetCommit for 1 partition(s): cgrp auto commit timer: returned: Success
%7|1610626632.654|HEARTBEAT|rdkafka#consumer-1| [thrd:main]: GroupCoordinator/8: Heartbeat for group "my_consumer_unique_id" generation id 9
%7|1610626634.676|OFFSET|rdkafka#consumer-1| [thrd:main]: Topic my_test_topic [0]: stored offset 25, committed offset 25: not including in commit
%7|1610626634.676|COMMIT|rdkafka#consumer-1| [thrd:main]: OffsetCommit for 1 partition(s): cgrp auto commit timer: returned: Local: No offset stored
%7|1610626635.673|HEARTBEAT|rdkafka#consumer-1| [thrd:main]: GroupCoordinator/8: Heartbeat for group "my_consumer_unique_id" generation id 9
%7|1610626638.696|HEARTBEAT|rdkafka#consumer-1| [thrd:main]: GroupCoordinator/8: Heartbeat for group "my_consumer_unique_id" generation id 9
%7|1610626639.680|OFFSET|rdkafka#consumer-1| [thrd:main]: Topic my_test_topic [0]: stored offset 25, committed offset 25: not including in commit
%7|1610626639.680|COMMIT|rdkafka#consumer-1| [thrd:main]: OffsetCommit for 1 partition(s): cgrp auto commit timer: returned: Local: No offset stored
%7|1610626641.713|HEARTBEAT|rdkafka#consumer-1| [thrd:main]: GroupCoordinator/8: Heartbeat for group "my_consumer_unique_id" generation id 9
%7|1610626644.684|OFFSET|rdkafka#consumer-1| [thrd:main]: Topic my_test_topic [0]: stored offset 25, committed offset 25: not including in commit
%7|1610626644.684|COMMIT|rdkafka#consumer-1| [thrd:main]: OffsetCommit for 1 partition(s): cgrp auto commit timer: returned: Local: No offset stored
%7|1610626644.729|HEARTBEAT|rdkafka#consumer-1| [thrd:main]: GroupCoordinator/8: Heartbeat for group "my_consumer_unique_id" generation id 9
%7|1610626647.752|HEARTBEAT|rdkafka#consumer-1| [thrd:main]: GroupCoordinator/8: Heartbeat for group "my_consumer_unique_id" generation id 9
%7|1610626649.687|OFFSET|rdkafka#consumer-1| [thrd:main]: Topic my_test_topic [0]: stored offset 25, committed offset 25: not including in commit
%7|1610626649.687|COMMIT|rdkafka#consumer-1| [thrd:main]: OffsetCommit for 1 partition(s): cgrp auto commit timer: returned: Local: No offset stored
%7|1610626650.772|HEARTBEAT|rdkafka#consumer-1| [thrd:main]: GroupCoordinator/8: Heartbeat for group "my_consumer_unique_id" generation id 9
%7|1610626653.787|HEARTBEAT|rdkafka#consumer-1| [thrd:main]: GroupCoordinator/8: Heartbeat for group "my_consumer_unique_id" generation id 9
%7|1610626654.691|OFFSET|rdkafka#consumer-1| [thrd:main]: Topic my_test_topic [0]: stored offset 25, committed offset 25: not including in commit
%7|1610626654.691|COMMIT|rdkafka#consumer-1| [thrd:main]: OffsetCommit for 1 partition(s): cgrp auto commit timer: returned: Local: No offset stored
end process
%7|1610626656.800|HEARTBEAT|rdkafka#consumer-1| [thrd:main]: GroupCoordinator/8: Heartbeat for group "my_consumer_unique_id" generation id 9
%7|1610626659.695|OFFSET|rdkafka#consumer-1| [thrd:main]: Topic my_test_topic [0]: stored offset 25, committed offset 25: not including in commit
%7|1610626659.700|COMMIT|rdkafka#consumer-1| [thrd:main]: OffsetCommit for 1 partition(s): cgrp auto commit timer: returned: Local: No offset stored

```

Also, Not sure about the meaning of these lines. 
```
Topic my_test_topic [0]: stored offset 25, committed offset 25: not including in commit
%7|1610626659.700|COMMIT|rdkafka#consumer-1| [thrd:main]: OffsetCommit for 1 partition(s): cgrp auto commit timer: returned: Local: No offset stored
```

When I restart the consumer I got 

```
1610627530.403|OFFSET|rdkafka#consumer-1| [thrd:main]: Topic t my_test_topic [0]: stored offset -1001, committed offset 25: not including in commit
%7|1610627530.403|COMMIT|rdkafka#consumer-1| [thrd:main]: OffsetCommit for 1 partition(s): cgrp auto commit timer: returned: Local: No offset stored
```

Then, if I also change the max.poll.interval.ms': 6000, that I agree it's too small, then I got the error message
```
consumer_conf = {'session.timeout.ms': 6000,
                         'max.poll.interval.ms': 6000,
                         'enable.auto.commit': 'true',
                         'log.connection.close': 'false',
                         'default.topic.config': {'auto.offset.reset': 'latest'}}
```

```
[2021-01-14T14:11:58+00:00] EXCEPTION KafkaError{code=_MAX_POLL_EXCEEDED,val=-147,str="Application maximum poll interval (6000ms) exceeded by 273ms"} on the consumer for topic(s) ['my_test_topic']
[2021-01-14T14:11:58+00:00] CLOSING consumer <confluent_kafka.avro.AvroConsumer object at 0x10bbf4048> with topic(s): ['my_test_topic']
%7|1610633518.685|CGRPOP|rdkafka#consumer-1| [thrd:main]: Group "my_consumer_unique_id" received op TERMINATE (v0) in state up (join state wait-revoke-rebalance_cb, v4 vs 0)
```

then, as you suggested, I remove the `raise Exception` from the code and the consumer does keep running 
```
long process
%7|1610634030.325|HEARTBEAT|rdkafka#consumer-1| [thrd:main]: GroupCoordinator/8: Heartbeat for group "my_consumer_unique_id" generation id 22
%7|1610634032.304|OFFSET|rdkafka#consumer-1| [thrd:main]: Topic my_test_topic [0]: stored offset 30, committed offset 29: setting stored offset 30 for commit
%7|1610634032.522|COMMIT|rdkafka#consumer-1| [thrd:main]: GroupCoordinator/8: OffsetCommit for 1 partition(s): cgrp auto commit timer: returned: Success
%7|1610634033.335|HEARTBEAT|rdkafka#consumer-1| [thrd:main]: GroupCoordinator/8: Heartbeat for group "my_consumer_unique_id" generation id 22
%4|1610634034.520|MAXPOLL|rdkafka#consumer-1| [thrd:main]: Application maximum poll interval (6000ms) exceeded by 55ms (adjust max.poll.interval.ms for long-running message processing): leaving group
%7|1610634034.520|LEAVE|rdkafka#consumer-1| [thrd:main]: Group "my_consumer_unique_id": leave (in state up)
%7|1610634034.520|MEMBERID|rdkafka#consumer-1| [thrd:main]: Group "my_consumer_unique_id": updating member id "rdkafka-3b109b12-25df-4049-ae08-8cce7c4decee" -> ""
%7|1610634034.520|REBALANCE|rdkafka#consumer-1| [thrd:main]: Group "my_consumer_unique_id" is rebalancing in state up (join-state started) with assignment: max.poll.interval.ms exceeded
%7|1610634034.520|ASSIGN|rdkafka#consumer-1| [thrd:main]: Group "my_consumer_unique_id": delegating revoke of 1 partition(s) to application rebalance callback on queue rd_kafka_cgrp_new: max.poll.interval.ms exceeded
%7|1610634034.520|CGRPJOINSTATE|rdkafka#consumer-1| [thrd:main]: Group "my_consumer_unique_id" changed join state started -> wait-revoke-rebalance_cb (v4, state up)
%7|1610634034.885|LEAVEGROUP|rdkafka#consumer-1| [thrd:main]:a LeaveGroup response received in state up
%7|1610634037.306|OFFSET|rdkafka#consumer-1| [thrd:main]: Topic my_test_topic [0]: stored offset 30, committed offset 30: not including in commit
%7|1610634037.307|COMMIT|rdkafka#consumer-1| [thrd:main]: OffsetCommit for 1 partition(s): cgrp auto commit timer: returned: Local: No offset stored
%7|1610634037.307|UNASSIGN|rdkafka#consumer-1| [thrd:main]: Unassign not done yet (0 wait_unassign, 1 assigned, 0 wait commit, F_WAIT_UNASSIGN, join state wait-revoke-rebalance_cb): OffsetCommit done (__NO_OFFSET)
%7|1610634042.311|OFFSET|rdkafka#consumer-1| [thrd:main]: Topic my_test_topic [0]: stored offset 30, committed offset 30: not including in commit
%7|1610634042.311|COMMIT|rdkafka#consumer-1| [thrd:main]: OffsetCommit for 1 partition(s): cgrp auto commit timer: returned: Local: No offset stored
%7|1610634042.311|UNASSIGN|rdkafka#consumer-1| [thrd:main]: Unassign not done yet (0 wait_unassign, 1 assigned, 0 wait commit, F_WAIT_UNASSIGN, join state wait-revoke-rebalance_cb): OffsetCommit done (__NO_OFFSET)
%7|1610634047.314|OFFSET|rdkafka#consumer-1| [thrd:main]: Topic my_test_topic [0]: stored offset 30, committed offset 30: not including in commit
%7|1610634047.314|COMMIT|rdkafka#consumer-1| [thrd:main]: OffsetCommit for 1 partition(s): cgrp auto commit timer: returned: Local: No offset stored
%7|1610634047.314|UNASSIGN|rdkafka#consumer-1| [thrd:main]: Unassign not done yet (0 wait_unassign, 1 assigned, 0 wait commit, F_WAIT_UNASSIGN, join state wait-revoke-rebalance_cb): OffsetCommit done (__NO_OFFSET)
%7|1610634052.318|OFFSET|rdkafka#consumer-1| [thrd:main]: Topic my_test_topic [0]: stored offset 30, committed offset 30: not including in commit
%7|1610634052.318|COMMIT|rdkafka#consumer-1| [thrd:main]: OffsetCommit for 1 partition(s): cgrp auto commit timer: returned: Local: No offset stored
%7|1610634052.318|UNASSIGN|rdkafka#consumer-1| [thrd:main]: Unassign not done yet (0 wait_unassign, 1 assigned, 0 wait commit, F_WAIT_UNASSIGN, join state wait-revoke-rebalance_cb): OffsetCommit done (__NO_OFFSET)
%7|1610634057.322|OFFSET|rdkafka#consumer-1| [thrd:main]: Topic my_test_topic [0]: stored offset 30, committed offset 30: not including in commit
%7|1610634057.322|COMMIT|rdkafka#consumer-1| [thrd:main]: OffsetCommit for 1 partition(s): cgrp auto commit timer: returned: Local: No offset stored
%7|1610634057.322|UNASSIGN|rdkafka#consumer-1| [thrd:main]: Unassign not done yet (0 wait_unassign, 1 assigned, 0 wait commit, F_WAIT_UNASSIGN, join state wait-revoke-rebalance_cb): OffsetCommit done (__NO_OFFSET)
%7|1610634062.326|OFFSET|rdkafka#consumer-1| [thrd:main]: Topic my_test_topic [0]: stored offset 30, committed offset 30: not including in commit
%7|1610634062.326|COMMIT|rdkafka#consumer-1| [thrd:main]: OffsetCommit for 1 partition(s): cgrp auto commit timer: returned: Local: No offset stored
%7|1610634062.326|UNASSIGN|rdkafka#consumer-1| [thrd:main]: Unassign not done yet (0 wait_unassign, 1 assigned, 0 wait commit, F_WAIT_UNASSIGN, join state wait-revoke-rebalance_cb): OffsetCommit done (__NO_OFFSET)
%7|1610634067.329|OFFSET|rdkafka#consumer-1| [thrd:main]: Topic my_test_topic [0]: stored offset 30, committed offset 30: not including in commit
%7|1610634067.329|COMMIT|rdkafka#consumer-1| [thrd:main]: OffsetCommit for 1 partition(s): cgrp auto commit timer: returned: Local: No offset stored
%7|1610634067.329|UNASSIGN|rdkafka#consumer-1| [thrd:main]: Unassign not done yet (0 wait_unassign, 1 assigned, 0 wait commit, F_WAIT_UNASSIGN, join state wait-revoke-rebalance_cb): OffsetCommit done (__NO_OFFSET)
%7|1610634072.330|OFFSET|rdkafka#consumer-1| [thrd:main]: Topic my_test_topic [0]: stored offset 30, committed offset 30: not including in commit
%7|1610634072.330|COMMIT|rdkafka#consumer-1| [thrd:main]: OffsetCommit for 1 partition(s): cgrp auto commit timer: returned: Local: No offset stored
%7|1610634072.330|UNASSIGN|rdkafka#consumer-1| [thrd:main]: Unassign not done yet (0 wait_unassign, 1 assigned, 0 wait commit, F_WAIT_UNASSIGN, join state wait-revoke-rebalance_cb): OffsetCommit done (__NO_OFFSET)
%7|1610634077.330|OFFSET|rdkafka#consumer-1| [thrd:main]: Topic my_test_topic [0]: stored offset 30, committed offset 30: not including in commit
%7|1610634077.330|COMMIT|rdkafka#consumer-1| [thrd:main]: OffsetCommit for 1 partition(s): cgrp auto commit timer: returned: Local: No offset stored
%7|1610634077.330|UNASSIGN|rdkafka#consumer-1| [thrd:main]: Unassign not done yet (0 wait_unassign, 1 assigned, 0 wait commit, F_WAIT_UNASSIGN, join state wait-revoke-rebalance_cb): OffsetCommit done (__NO_OFFSET)
%7|1610634082.333|OFFSET|rdkafka#consumer-1| [thrd:main]: Topic my_test_topic [0]: stored offset 30, committed offset 30: not including in commit
%7|1610634082.333|COMMIT|rdkafka#consumer-1| [thrd:main]: OffsetCommit for 1 partition(s): cgrp auto commit timer: returned: Local: No offset stored
%7|1610634082.333|UNASSIGN|rdkafka#consumer-1| [thrd:main]: Unassign not done yet (0 wait_unassign, 1 assigned, 0 wait commit, F_WAIT_UNASSIGN, join state wait-revoke-rebalance_cb): OffsetCommit done (__NO_OFFSET)
%7|1610634087.337|OFFSET|rdkafka#consumer-1| [thrd:main]: Topic my_test_topic [0]: stored offset 30, committed offset 30: not including in commit
%7|1610634087.337|COMMIT|rdkafka#consumer-1| [thrd:main]: OffsetCommit for 1 partition(s): cgrp auto commit timer: returned: Local: No offset stored
%7|1610634087.337|UNASSIGN|rdkafka#consumer-1| [thrd:main]: Unassign not done yet (0 wait_unassign, 1 assigned, 0 wait commit, F_WAIT_UNASSIGN, join state wait-revoke-rebalance_cb): OffsetCommit done (__NO_OFFSET)
end process
[2021-01-14T14:21:29+00:00] Consumer Error KafkaError{code=_MAX_POLL_EXCEEDED,val=-147,str="Application maximum poll interval (6000ms) exceeded by 55ms"}
%7|1610634089.665|CGRPOP|rdkafka#consumer-1| [thrd:main]: Group "my_consumer_unique_id" received op ASSIGN (v0) in state up (join state wait-revoke-rebalance_cb, v4 vs 0)
%7|1610634089.665|ASSIGN|rdkafka#consumer-1| [thrd:main]: Group "my_consumer_unique_id": new assignment of 0 partition(s) in join state wait-revoke-rebalance_cb
```

In the end, 

1. When I first got the message, I was using the default value for  `max.poll.interval.ms` that I know is 300000ms and `'session.timeout.ms': 6000,` and no blocking function to process the incoming message. I don't understand why it occurred. 
2. if I use a blocking function for 30sec with `'session.timeout.ms': 6000` I didn't get the error message. 
3. If I also set the `max.poll.interval.ms': 6000` and use the same blocking function, then I get the error message. 
4. If I remove the `raise` from the code, the consumer stays alive even if it does not process new messages till the blocking function returns, but it's expected.

**Comment by richecr** [comment]
Any news on this, I'm getting the same problem here

**Comment by richecr** [comment]
any news about this? I'm getting the same problem here

It has been corrected? I'm using the `confluent-kafka = "==2.0.2"`

**Comment by ChanneRen** [comment]
> It is not exception, it is a log message, and it can't and shouldn't be catched. The max.poll.interval.ms is there for a reason; it let's you specify how long your consumer owns the assigned partitions following a rebalance - doing anything with the data when this period has expired means there might be duplicate processing. So the solution is to either: a. indicate that your application is still alive by calling poll() - if you dont want more messages you will need to pause() your partitions first (but do note that this comes at the cost of purging the pre-fetch queue). b. increase max.poll.interval.ms to your maximum http retry time.

@edenhill thanks for your sharing. But I still have a question on recovery. When this issue happens, the consumer will rejoin when it executes `poll` again. I wonder if this only works for high-level consumers. We use librdkafka by low-level mode, and we found the consumer could not recover event it executed `poll` again. It polled nothing and
